### PR TITLE
add Learn the concepts --> Peer-to-Peer Systems and Multi-Agent Systems

### PR DESCRIPTION
**Pull request by FelixNicolaeBucsa**
Hey there. This is the PR for "_Peer-to-Peer systems_" and "_Agent-based and Multi-Agents based systems_" topics in the Learn the Concepts section. Please have a view at it when possible @5A11 @GataKamsky @StephenBrownFetch @VijaySharma815 @ejfitzgerald

### CODE REVIEW

**Comment by VijaySharma815** [comment]
learn the concept


@FelixNicolaeBucsa can't we make it in 2-3 paragraphs? actually reading that much big para user can get bored.

**Comment by VijaySharma815** [comment]
Suggestion:

It would be great if we can add some graphics.

**Comment by GataKamsky** [comment]
I agree with @VijaySharma815 's comments. Would be good to break it into multiple paras and add some pictures.

That being said, we can also push this version out and modify it in the next version.

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+As we can see, MAS have found multiple applications including _modeling complex systems_, _smart grids_ and _computer networks_. The main application domains of MAS are _ambient intelligence_, _grid computing_, _electronic business_, _the semantic web_, _bioinformatics and computational biology_, _monitoring and control_, _resource management_, _education_, _space_, _military_ and _manufacturing applications_, and so on.
+
+**Despite their wide applicability, there are still a number of challenges faced by MAS including coordination between agents, security, and task allocation.**
```
Comment (VijaySharma815) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]: there are still a number of challenges faced by MAS including coordination between agents -> there are still several challenges faced by MAS including coordination between agents

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+MAS have received tremendous attention from scholars in different disciplines, ranging from computer science to civil engineering, in order to solve complex problems by subdividing them into smaller tasks. The individual tasks are allocated to autonomous agents, and each one of them decides on a proper action to solve the task using multiple inputs (e.g. history of actions, interactions with its neighboring agents, and its goal).
+
+A good example of a MAS is represented by _smart-grids_, in which different agents work together collaboratively, but independently and in a self-interested manner. Another important application of such systems is found in _computer gaming_ and _computer simulations_. Such fields increasingly adopted MASs. Here, each participant has its own agent and interacts in the environment populated also by other agents. _Self-driving vehicles_ are another great example of an agent based system. In this respect, each vehicle acts as an agent, directed by their user and acting independently from other agents within a set of rules, such as not exceeding the speed limits, and responsive to the environment around them.
```
Comment (VijaySharma815) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]: responsive to the environment around them -> being responsive to the environment around them.

**Review by VijaySharma815 — COMMENTED** [review]
overall content looks good to me!

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# Agent-Based Systems
+
+An **agent** is a computer program that performs various actions _continuously_ and _autonomously_  (with limited or no interference) on behalf of an individual or an organization. Agents have behaviors, often described by simple rules, and interactions with other agents, which in turn influence their behaviors. The most important defining characteristic of an agent is its capability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviors that allow them to make independent decisions. Typically, agents are active, initiating their actions to achieve their internal goals, rather than merely passive, reactively responding to other agents and the environment.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 

The most important defining characteristic of an agent is its ability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 

What differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that they are _reactive_, they respond to other agents and changes in their environments, and also _proactive_, meaning they take the initiative and perform actions to achieve their goals.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# Agent-Based Systems
+
+An **agent** is a computer program that performs various actions _continuously_ and _autonomously_  (with limited or no interference) on behalf of an individual or an organization. Agents have behaviors, often described by simple rules, and interactions with other agents, which in turn influence their behaviors. The most important defining characteristic of an agent is its capability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviors that allow them to make independent decisions. Typically, agents are active, initiating their actions to achieve their internal goals, rather than merely passive, reactively responding to other agents and the environment. 
+
+Agents have **five key characteristics**:
+
+1. **Representation**: The first key characteristic of agents is that they operate on behalf of an individual or organization. Agents are always operating for an entity.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
1. **Representation**: Agents are owned by and operate on behalf of an entity, for example an individual, family, company, government, or object, and look after their owner's interests.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+An **agent** is a computer program that performs various actions _continuously_ and _autonomously_  (with limited or no interference) on behalf of an individual or an organization. Agents have behaviors, often described by simple rules, and interactions with other agents, which in turn influence their behaviors. The most important defining characteristic of an agent is its capability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviors that allow them to make independent decisions. Typically, agents are active, initiating their actions to achieve their internal goals, rather than merely passive, reactively responding to other agents and the environment. 
+
+Agents have **five key characteristics**:
+
+1. **Representation**: The first key characteristic of agents is that they operate on behalf of an individual or organization. Agents are always operating for an entity. 
+
+2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to the rules they have been given.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to their internal reasoning system.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+3. **Self-interested**: They operate in their own interests and not necessarily in the interests of other agents.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
3. **Self-interested**: Each agent primarily looks after its own interests (which is aligned with those of its owner) and not necessarily the interests of other agents.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+4. **Goal-oriented**:  agents have goals to achieve with respect to their behaviors. This allows an agent to compare the outcome of its behaviors relative to its goals and adjust its responses and behaviors in future interactions.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
4. **Proactive**:  Agents have goals to achieve allowing them to take the initiative and perform actions that get them closer to achieving their goals. This means an agent typically compares the outcome of different actions relative to its goals and selects the one that takes it closer to them. 
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+5. **Reactive**: Agents are reactive. This is because they react to the circumstances around them and act consequently. A great example of this is an automatic thermostat, that turns the heating on or off when the temperature changes.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is a heating system with thermostat that constantly monitors the environment and turns the heating on or off when the temperature changes.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
+# Agent-Based Systems
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
## Agent
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and acts _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
+# Agent-Based Systems
+
+An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 
+
+The most important defining characteristic of an agent is its ability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
The most important defining characteristic of an agent is its autonomy; that is, to act on its own without external direction from its owner in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 12]
```diff
@@ -0,0 +1,56 @@
+# Agent-Based Systems
+
+An **agent** is a piece of software that _represents_ an entity (individual, organisation, or object) and performs various actions _continuously_ and _autonomously_  (with limited or no interference) on their behalf. 
+
+The most important defining characteristic of an agent is its ability to act autonomously, that is, to act on its own without external direction in response to situations it encounters. Agents are endowed with behaviours that allow them to make independent decisions. 
+What differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that they are _reactive_, they respond to other agents and changes in their environments, and also _proactive_, meaning they take the initiative and perform actions to achieve their goals.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
What differentiates agents with other software paradigms (e.g. smart contracts, web apps) is that as well as being _reactive_, meaning they respond to other agents and changes in their environments, they are also _proactive_, which means that they take the initiative and perform actions to achieve their goals.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+Agents have **five key characteristics**:
+
+1. **Representation**: Agents are owned by and operate on behalf of an entity, for example an individual, family, company, government, or object, and look after their owner's interests. 
+
+2. **Autonomy**: Agents operate with limited or no interference, they do not need to be told what to do constantly but perform actions continuously according to their internal reasoning system.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
2. **Autonomy**: Agents operate with limited or no interference and do not need to be constantly told what to do. They perform actions continuously according to their internal reasoning system.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+3. **Self-interested**: Each agent primarily looks after its own interests (which is aligned with those of its owner) and not necessarily the interests of other agents.
+
+4. **Proactive**:  Agents have goals to achieve allowing them to take the initiative and perform actions that get them closer to achieving their goals. This means an agent typically compares the outcome of different actions relative to its goals and selects the one that takes it closer to them.
+
+5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is a heating system with thermostat that constantly monitors the environment and turns the heating on or off when the temperature changes.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
5. **Reactive**: Agents respond to other agents, services, etc., and changes in their environments. A great example of this behaviour is a heating system with a thermostat that constantly monitors its environment and turns the heating on or off when the temperature changes.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+_Agent characteristics and behaviors may vary in their extent and sophistication_. An agent may differ from other agents, for instance, in the amount of information considered in the agent’s decisions process, the agent’s internal models of the external world, the agent’s view of the possible reactions of other agents in response to its actions, and the extent of memory of past events the agent retains and uses in making its decisions. Agents may also be endowed with  different amounts of resources or accumulate different levels of resources as a result of agent interactions, further differentiating agents.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
_Agent characteristics and behaviors may vary in their extent and sophistication_. An agent may differ from other agents, for instance, in the amount of information considered in the agent’s decisions process, its internal models of the external world, its view of the possible reactions of other agents in response to its actions, and the size of the memory of past events the agent retains and uses in making its decisions. Further differentiating factors are the amounts of resources used by agents or accumulated as a result of their interactions.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+Agent-based modeling has been used in an enormous variety of applications spanning the physical, biological, social, and management sciences. In the biological sciences, agent-based modeling is used to model cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets (e.g. agent-based simulation to model possible futures for a market directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market).
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]:
```suggestion
## Agent-Based Modelling

Agent-based modelling has been used in enormous kinds of applications spanning the physical, biological, social, and management sciences. In biological sciences, agent-based modelling is used to simulate cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). 

Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets. For example, agent-based simulation is is used to model possible futures for a market that is directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market.
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+As it is possible to think, there are a growing number of agent-based applications in a variety of fields and disciplines directed towards the study of a large number of real-world applications. Examples of agent-based systems, for which large-scale agent-based models have been developed, include: traffic, air traffic control, military command and control, physical infrastructures and markets, such as electric power and integrated energy markets.
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md:31]:
```suggestion
Unsurprisingly, there are a growing number of agent-based applications in a variety of fields and disciplines that focus on the study of large numbers of real-world phenomena. Some other examples of systems, for which large-scale agent-based models have been developed, include: traffic, air traffic control, military command and control, physical infrastructures and markets such as electric power and integrated energy markets. 
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md (hunk 16 above)
Comment (5A11) [reply docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]: Add a paragraph at the beginning of agent-based modelling introducing introducing what it is, what it's aim and scope are, etc.

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+As it is possible to think, there are a growing number of agent-based applications in a variety of fields and disciplines directed towards the study of a large number of real-world applications. Examples of agent-based systems, for which large-scale agent-based models have been developed, include: traffic, air traffic control, military command and control, physical infrastructures and markets, such as electric power and integrated energy markets. 
+
+# Multi-Agents Systems
```
Comment (5A11) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md:33]:
```suggestion
## Multi-Agents Systems
```

File: mkdocs.yml [hunk 19]
```diff
@@ -38,6 +38,9 @@ theme:
 nav:
   - Home : "index.md"
   - Bug Bounty Program: "bug_bounty.md"
+  - Learn the Concepts:
+    - Peer-to-Peer Applications: 'learn_the_concepts/peer_to_peer_systems.md'
+    - Agent-based and Multi-agents based systems: 'learn_the_concepts/agent-based_and_multi-agents_systems.md'
```
Comment (5A11) [inline mkdocs.yml:43]:
```suggestion
    - Agent-based and Multi-Agents Systems: 'learn_the_concepts/agent-based_and_multi-agents_systems.md'
```

File: docs/learn_the_concepts/agent-based_and_multi-agents_systems.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+Agent-based modelling has been used in enormous kinds of applications spanning the physical, biological, social, and management sciences. In biological sciences, agent-based modelling is used to simulate cell behavior and interaction, the workings of the immune system, tissue growth, and disease processes. Archaeology and anthropology make use of large scale agent-based modeling by providing a virtual environment for long-vanished civilizations. In ecology, agent-based modeling is used to model diverse populations of individuals and their interactions (e.g. Agent-based epidemic and pandemic models). 
+
+Agent-based models have also been developed in the fields of economics, sociology, anthropology, and cognitive science. Various social phenomena have been investigated using agent-based models that are not easily modeled using other approaches. Here, agent-based models are being used to analyze existing and hypothetical markets. For example, agent-based simulation is is used to model possible futures for a market that is directed towards space tourism, to analyze how companies represented by agents would compete and offer products to customers in this hypothetical market.
```
Comment (VijaySharma815) [inline docs/learn_the_concepts/agent-based_and_multi-agents_systems.md]: For example, agent-based simulation is is used to model possible futures for a market -> For example, agent-based simulation is used to model possible futures for a market.

**Review by VijaySharma815 — APPROVED** [review]

File: docs/learn_the_concepts/peer_to_peer_systems.md [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+1. **No middlemen needed**: One of the key characteristics of peer-to-peer systems is that middlemen are not always required, though they are allowed. For instance, when booking a hotel room with a travel agent, the users will be going through a middleman but if the user books a hotel room directly with the hotel itself, then this depicts a direct communication, and hence, they are communicating on a peer-to-peer basis.
+
+2. **Direct links**: direct communication between both parties can be established only in the presence of direct links between them.
+
+3. **Direct interaction**: It is not enough just for parties to be able to communicate directly, they have to be able to actually interact. Peer-to-peer systems are characterized by direct interaction. In the hotel room booking example, making a booking with the hotel with the hotel sending back a booking confirmation actually consists of a direct interaction between the two parties.
```
Comment (VijaySharma815) [inline docs/learn_the_concepts/peer_to_peer_systems.md]: making a booking with the hotel with the hotel sending back a booking -> making a booking with the hotel sending back a booking

File: docs/learn_the_concepts/peer_to_peer_systems.md [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+Another example of P2P networking is given by some online gaming platforms which adopt  a P2P structure for downloading games between users. The company [Blizzard Entertainment](https://www.blizzard.com/en-gb/) distributed Diablo III, StarCraft II, and World of Warcraft using P2P. Another large publisher, Wargaming, does the same with their World of Tanks, World of Warships, and World of Warplanes games.
```
Comment (VijaySharma815) [inline docs/learn_the_concepts/peer_to_peer_systems.md]: platforms which adopt  a P2P structure for downloading games between users -> platforms which adopt a P2P structure for downloading games between users

**Pull request merged**